### PR TITLE
Calibration

### PR DESCRIPTION
**Pull request by mtereshkin**

### CODE REVIEW

**Comment by mephody-bro** [comment]
Как стартовая точка выглядит хорошо, но надо будет дорабатывать.
1. Сначала более формальное:
Просмотри pep8 (https://www.python.org/dev/peps/pep-0008/). Это общий гид по стилю от сообщества. В частности, название методов в классе должно быть маленькими буквами в snake_case. Единный стиль в целом упрощает чтение кода при работе с другими людьми.

2. Подразумевается, что скриптовая часть -- это теперь часть библиотеки. Поэтому она должна лежать в alpaca и импорироваться как alpaca.calibrator
Тут есть практический момент - в примерах обычно подразумевается, что библиотека установлено именно как библиотека, а не как локальные файлы. Чтобы эмулировать это и при этом подхватывались изменения, нужно из локальной папки установаить в editable режиме. Просто из корня альпаки набери
```
pip install -e .
```

3. Про мнист - его можно подгружать без необходимости класть .csv файлы (в sklearn есть fetch_mldata, у нас есть свой подгрузчик, посмотри как в classification_uq.ipynb он подгружается). Сейчас просто выходит, что пример не работает без того, чтобы где-то искать эти csv, что плохо.

4. Почему у тебя отдельный compute_ece есть в ноутбуке и в скрипте?

5. Понятно, что fit/predict у sklearn удобный, но по факту все сейчас сеточки пишут на торче. Нужно адаптировать под pytorch модели.

6. Всё класть в один класс -- это будет не слишком хорошо адапатироваться. Понятно, что сейчас по сути ты делаешь из класса калибратора модуль, где всё сразу есть. Но кажется, что всякие метрики лучше выносить в независимые отдельные функции., которым ты явно там даёшь данные, так они более "чистые", их легче тестировать и переиспользовать.

Для использования попимо бенчмарков, круто, чтобы они заворачивались в классы обёртки, вроде как делается вот здесь https://github.com/gpleiss/temperature_scaling. Мы для всех методов так сделать можем?

Т.е. идеально чтобы в модуле были бы:
- Классы-обёртки под каждый метод калибровки
- Функции, считающие метрики
- Какие-то ещё вспомогательные функции, которые ты внутри используешь. Если не предполагается, что их будут использовать вне модуля, начинай имя с подчёркивания, например def _choose_data()

**Comment by mephody-bro** [comment]
Круто, сильно продвинулись в нужную сторону.
1. Можно, чтобы интерфейс был похож у трёх классов? Типо, везде scaling, а не matrix_scaling и vector_scaling. (это как в торче и в полносвязной, и в свёрточной сети метод называется forward, хотя операция разная)
2. В ноутбуке строчки с подсчётом метрик дублируются, вынеси в функцию.
3. Ещё, ты просто через upload делаешь, а не pull/push? Попробуй как я описывал https://hackmd.io/@mephody-bro/B10bgnlkw#/. Более осмысленные commit сообщения, которые кратко описывают изменения, могут облегчить жизнь. Плюс с репозиторием ты можешь откатиться, если что-то пойдёт не так. Да и если будут изменения в 10 файлах, а не двух, то сильно тебе жизнь облегчит.

**Comment by mtereshkin** [comment]
1. Поменяли имена методов  на  scaling;
2. Добавили функцию в ноутбуке, которая считает все метрики и выводит их.
3. Вроде получилось нормально сделать коммиты через git

**Comment by mephody-bro** [comment]
Ок, отлично. И ещё одну итерацию, более косметическую.
1. Нужно, чтобы для функций ошибок порядок аргументов был последовательный. Сейчас labels и probs местами меняются постоянно
2. Используйте для всех обёрток в примере метод forward вместо ручного scaling_logits + сделайте, чтобы он софтмаксимл сам. Есть, к слову, torch.softmax
3. Для scaling передавайте логиты и лейблы как аргументы, а не через конструктор. Аналогично, как в sklearn параметры для fit передаются как аргументы, а не как конструктор. Если нужны доп параметры, типо размера слоя в матрицы, то передавайте их.
4. Немножко docstring для классов было бы полезно.
